0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+  `attached_balance + deficit - burnt_balance` back to the `account_balance` and do not issue a refund;
+* If `deficit <= 0` (we have deposited to the account more than what we have withdrawn) we withdraw `abs(deficit)`
+  from the `account_balance` and issue a refund of `attached_balance - burnt_balance`.
+  
+Note, for both cases above, `burnt_balanace` is never greater than `attached_balance + deficit` at any point in the program. This
+might require a formal proof later.
+
+---
+The following spec is the same for all functions:
+```
+account_balance(balance_ptr: u64)
+attached_balance(balance_ptr: u64)
+remaining_balance(balance_ptr: u64)
+burnt_balance(balance_ptr: u64)
+```
+ -- writes the value into the `u128` variable pointed by `balance_ptr`.
+
+###### Panics
+* If `balance_ptr + 16` points outside the memory of the guest with `MemoryAccessViolation`;
+
+###### Current bugs
+* Use a different name;
+* Unclear how they behave with the callbacks.
+
+---
+```rust
+deposit(min_amount_ptr: u64, max_amount_ptr: u64, result_ptr: u64)
+```
+Moves balance from `attached_balance` to `account_balance`.
+If `min_amount <= attached_balance` moves `min(max_amount, max(min_amount, attached_balance))`. The moved amount is
+saved into `u128` variable pointed by `result_ptr`. `min_amount_ptr` and `max_amount_ptr` point to `u128` variables.
+
+###### Panics
+* If `min_amount_ptr + 16` or `max_amount_ptr + 16` or `result_ptr + 16` point outside the memory of the guest with `MemoryAccessViolation`;
+
+---
+
+```rust
+withdraw(min_amount_ptr: u64, max_amount_ptr: u64, result_ptr: u64)
+```
+Moves balance from `account_balance` to `attached_balance`.
+If `min_amount <= attached_balance` moves `min(max_amount, max(min_amount, account_balance))`. The moved amount is
+saved into `u128` variable pointed by `result_ptr`. `min_amount_ptr` and `max_amount_ptr` point to `u128` variables.
+
+###### Panics
+* If `min_amount_ptr + 16` or `max_amount_ptr + 16` or `result_ptr + 16` point outside the memory of the guest with `MemoryAccessViolation`;
+
+## Math
+
+```rust
+random_buf(buf_len: u64, buf_ptr: u64)
+```
+Writes random bytes in the given memory location on the guest. Does not
+work with scratch buffer.
+
+###### Panics
+* If `buf_len + buf_ptr` points outside the memory of the guest with `MemoryAccessViolation`;
+
+---
+```rust
+random_u64() -> u64
+``` 
+Returns a random `u64` variable.
+
+---
+```rust
+sha256(value_len: u64, value_ptr: u64, register_id: u64)
+```
+Hashes the random sequence of bytes using sha256 and returns it into `register_id`.
+###### Panics
+* If `value_len + value_ptr` points outside the memory or scratch buffer uses more memory than the limit with `MemoryAccessViolation`.
+
+###### Current bugs
+* Current name `hash` is not specific to what hash is being used.
+* We have `hash32` that largely duplicates the mechanics of `hash` because it returns the first 4 bytes only.
+
+---
+```rust
+check_ethash(block_number_ptr: u64,
```

Comment:
We should completely remove it until ethash bridge works

Reply:
I don't think so. This is needed for implementing the bridge

Reply:
Discussed with @ilblackdragon . We keep it for the team that will be working on it.